storage/flush.go

```diff
@@ -154,42 +129,46 @@ func (m *flushManager) RateLimitOptions() ratelimit.Options {
 	return m.pm.RateLimitOptions()
 }
 
-// flushTimes returns a list of times we need to flush data blocks for.
-func (m *flushManager) flushTimes(t time.Time) []time.Time {
-	earliest, latest := m.FlushTimeStart(t), m.FlushTimeEnd(t)
+func (m *flushManager) flushTimes(curr time.Time) []time.Time {
+	earliest, latest := m.FlushTimeStart(curr), m.FlushTimeEnd(curr)
 
 	// NB(xichen): could preallocate slice here.
 	var flushTimes []time.Time
-	m.RLock()
-	for flushTime := latest; !flushTime.Before(earliest); flushTime = flushTime.Add(-m.blockSize) {
-		if m.needsFlushWithLock(flushTime) {
-			flushTimes = append(flushTimes, flushTime)
+	for t := latest; !t.Before(earliest); t = t.Add(-m.blockSize) {
+		if m.needsFlush(t) {
+			flushTimes = append(flushTimes, t)
 		}
 	}
-	m.RUnlock()
 
 	return flushTimes
 }
 
-// needsFlushWithLock returns true if we need to flush data for a given time.
-func (m *flushManager) needsFlushWithLock(t time.Time) bool {
-	flushState, exists := m.flushStates[t]
-	if !exists {
-		return true
+// needsFlush returns true if we need to flush data for a given time.
+func (m *flushManager) needsFlush(t time.Time) bool {
+	maxRetries := m.opts.MaxFlushRetries()
+	namespaces := m.database.getOwnedNamespaces()
+	for _, n := range namespaces {
+		state := n.FlushState(t)
+		if state.Status != fileOpSuccess &&
```

Comment:
Might be worth moving this logic to `needsFlush` in the namespace implementation for better encapsulation